contributors/design-proposals/bulk_watch.md

```diff
@@ -0,0 +1,458 @@
+type Watch struct {
+	Selector ListSelector
+}
+
+// Request to stop watching objects from a watch identified by the channel.
+type CloseWatch struct {
+	Channel Identifier
+}
+
+// Depending on the request, channel that was created or deleted.
+type Response struct {
+	Channel Identifier
+}
+```
+With the following structure we can guarantee that we only send and receive
+objects of a single type over the channel.
+
+Another detail we need to point at is about semantic. If there are multiple
+selectors selecting exactly the same object, the object will be send multiple
+times (once for every channel interested in that object).
+If we decide to also have http-POST-based watch, since there will be basically
+a single channel there (as it is in watch in regular API), such an object will
+be send once. This semantic difference needs to be explicitly described.
+
+*TODO: Since those channels for individual watches are kind of independent,
+we need to decide whether they are supposed to be synchronized with each other.
+In particular, if we have two channels watching for objects of the same type
+are events guaranteed to be send in the increasing order of resource versions.
+I think this isn't necessary, but it needs to be explicit.*
+
+Yet another thing to consider is what if the server wants to close the watch
+even though user didn't request it (some potential reasons of it may be failed
+periodic ACL check or some kind of timeout).
+We will solve it by saying that in such situation we will send an error via
+the corresponding channel and every error automatically closes the channel.
+It is responsibility of a user to re-subscribe if that's possible.
+
+Since within a single Subsecribe request, we can ask only for objects of a
+single type, there is also no reason to introduce `BulkEvent` in this case.
+We can simply reuse `watch.Event` since client knows objects of what type
+will be send via a given channel.
+
+
+### Implementation details
+
+ The implementation of bulk get and bulk watch will work as an aggregator.
+Whenever a request is coming, it will:
+- check what resource types are requested in this request
+- for every resource type, combine only parts of the filter that are about
+this particular resource type and send exactly one request down the stack
+- gather responses for all those resource types and combine them into single
+response to the user
+
+ This has to support API aggregation as well. The same API (endpoint) has
+to be implemented in kube-aggregator, though its implementation is different.
+In the aggrgator, there are two alternatives:
+1. aggregator forwards the request to all apiservers and aggregates results
+2. based on discovery information, aggregator knows which type is supported
+by which apiserver, so it is forwarding requests with just appropriate
+resource types to apiservers and then aggregates results
+Neither of those is difficult, so we can probably proceed with the second.
```

Comment:
It depends how efficiently the other option is implemented. But potentially yes, you're right.